src/conversions/num_complex.rs

```diff
@@ -152,6 +152,19 @@ macro_rules! complex_conversion {
 
                 #[cfg(any(Py_LIMITED_API, PyPy))]
                 unsafe {
+                    use crate::{intern, type_object::PyTypeInfo};
+
+                    let obj = if PyComplex::is_type_of(obj) {
+                        obj
+                    } else if let Ok(method) =
+                        obj.get_type().getattr(intern!(obj.py(), "__complex__"))
+                    {
+                        method.call1((obj,))?
+                    } else {
+                        // This will likely result in an error in `PyComplex_RealAsDouble`, but
```

Comment:
I don't think this too helpful and would probably to just return the error of calling `__complex__` directly.

If I read https://docs.python.org/3/c-api/complex.html#c.PyComplex_AsCComplex correctly, we would also need to do more, e.g.

```rust
let obj = obj.is_instance_of::<PyComplex>() {
  obj
} else if let Ok(obj) = obj.method_call0(intern!(obj.py(), "__complex__")) {
  obj
} else {
  let num = obj.extract::<f64>()?;
  return Ok(Complex::from(num));
};
```

Reply:
Oh sorry, my comment is poor here: in your case, `PyComplex_RealAsDouble` does this conversion in the same way that `PyComplex_AsCComplex` does, so the code would continue on to succeed if it's a valid float.

Do you prefer me to change the comment or to switch to your implementation?

Reply:
Sounds good: done in b30169ae and aa1dcb04.